### PR TITLE
test rgw restore index tool

### PR DESCRIPTION
**Pull request by viduship**
test rgw restore index tool 

This would address the manual verification of many Bugzillas over the different product releases.

https://bugzilla.redhat.com/show_bug.cgi?id=2247138
https://bugzilla.redhat.com/show_bug.cgi?id=2240992
https://bugzilla.redhat.com/show_bug.cgi?id=2182385

JIRA ticket https://issues.redhat.com/browse/RHCEPHQE-12036

[POLARIAN test case : CEPH-83575473](https://polarion.engineering.redhat.com/polarion/#/project/CEPH/workitem?id=CEPH-83575473)

passed logs : http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-IAYSHB

### CODE REVIEW

File: rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+
+        bkt_stat_cmd = f"radosgw-admin bucket stats --bucket {bucket.name}"
+        restored_num_objects = json.loads(utils.exec_shell_cmd(bkt_stat_cmd))["usage"][
+            "rgw.main"
+        ]["num_objects"]
+
+        if restored_num_objects == num_objects:
+            log.info("restore bucket index is successful!")
+        else:
+            raise AssertionError("restore bucket index is successful!")
```
Comment (hmaheswa) [inline rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py:149]: @viduship please fix this error as "restore bucket index is not successful!"

File: rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        bucket_id = json.loads(utils.exec_shell_cmd(bkt_stat_cmd))["id"]
+
+        log.info(
+            f"the shards and number of objects before rados clearomap are {num_shard} shards and {num_objects} objects"
+        )
+        for shard in range(num_shard):
+            cmd_rados_clear_omap = f"rados clearomap -p {zone_name}.rgw.buckets.index  .dir.{bucket_id}.{bucket_gen}.{shard}"
+            utils.exec_shell_cmd(cmd_rados_clear_omap)
+
+        bkt_stat_cmd = f"radosgw-admin bucket stats --bucket {bucket.name}"
```
[the remark this answers is not in the document]
Comment (viduship) [reply rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py:128]: @hmaheswa , it is required to test bucket stats shows the bucket instances are cleared via the rados clearomap.

File: rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py (hunk 1 above)
Comment (viduship) [reply rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py:149]: done @hmaheswa

File: rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (viduship) [reply rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py:128]: http://magna002.ceph.redhat.com/cephci-jenkins/cephci-run-IAYSHB

**Review by hmaheswa — APPROVED** [review]

File: rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    io_info_initialize.initialize(basic_io_structure.initial())
+    ceph_conf = CephConfOp(ssh_con)
+    rgw_service = RGWService()
+    log.info("starting IO")
+    config.user_count = 1
+    user_info = s3lib.create_users(config.user_count)
+    user_info = user_info[0]
+    auth = Auth(user_info, ssh_con, ssl=config.ssl, haproxy=config.haproxy)
+    rgw_conn = auth.do_auth()
+    config.bucket_count = 1
```
[the remark this answers is not in the document]
Comment (viduship) [reply rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py:61]: done.

File: rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+        utils.exec_shell_cmd(restore_index_cmd)
+        time.sleep(60)
+        log.info(
+            "the number of objects after the restore-index tool should be same as before."
+        )
+
+        bkt_stat_cmd = f"radosgw-admin bucket stats --bucket {bucket.name}"
+        restored_num_objects = json.loads(utils.exec_shell_cmd(bkt_stat_cmd))["usage"][
+            "rgw.main"
+        ]["num_objects"]
```
[the remark this answers is not in the document]
Comment (viduship) [reply rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py:138]: done

File: rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py [hunk 5]
```diff
@@ -0,0 +1,204 @@
+"""
+test_rgw_restore_index_tool - Test resharding operations on bucket
+
+Usage: test_rgw_restore_index_tool.py -c <input_yaml>
+
+<input_yaml>
+    Note: any one of these yamls can be used
+    test_rgw_restore_index_multipart_uploads.yaml
```
[the remark this answers is not in the document]
Comment (viduship) [reply rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py]: @mkasturi18 , hey I plan to add it in the future. I have update with a comment.

File: rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+    test_rgw_restore_index_versioned_buckets.yaml
+
+Operation:
+    Create user and a bucket and enable bucket versioning
+    Upload ~300 objects to the bucket (with 3 versions of each object)
```
[the remark this answers is not in the document]
Comment (viduship) [reply rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py]: @mkasturi18 , yeah, thanks for reviewing. done

File: rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+        )
+        for shard in range(num_shard):
+            cmd_rados_clear_omap = f"rados clearomap -p {zone_name}.rgw.buckets.index  .dir.{bucket_id}.{bucket_gen}.{shard}"
+            utils.exec_shell_cmd(cmd_rados_clear_omap)
+
+        bkt_stat_cmd = f"radosgw-admin bucket stats --bucket {bucket.name}"
+        log.info(
+            f"bucket stats after clearing the bucket instance objects: {utils.exec_shell_cmd(bkt_stat_cmd)}"
+        )
+        log.info("Install the jq utility")
```
[the remark this answers is not in the document]
Comment (viduship) [reply rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py]: @mkasturi18 sure, I have added it in the https://github.com/red-hat-storage/cephci/pull/3213/files, please let me know if that good

File: rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+
+import v2.lib.resource_op as s3lib
+import v2.utils.utils as utils
+from v2.lib.exceptions import RGWBaseException, TestExecError
+from v2.lib.resource_op import Config
+from v2.lib.rgw_config_opts import CephConfOp, ConfigOpts
+from v2.lib.s3.auth import Auth
+from v2.lib.s3.write_io_info import BasicIOInfoStructure, BucketIoInfo, IOInfoInitialize
+from v2.tests.s3_swift import reusable
+from v2.tests.s3_swift.reusables import quota_management as quota_mgmt
```
[the remark this answers is not in the document]
Comment (viduship) [reply rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py]: sure

File: rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+def test_exec(config, ssh_con):
+    io_info_initialize = IOInfoInitialize()
+    basic_io_structure = BasicIOInfoStructure()
+    write_bucket_io_info = BucketIoInfo()
+    io_info_initialize.initialize(basic_io_structure.initial())
+    ceph_conf = CephConfOp(ssh_con)
+    rgw_service = RGWService()
+    log.info("starting IO")
+    config.user_count = 1
```
[the remark this answers is not in the document]
Comment (viduship) [reply rgw/v2/tests/s3_swift/test_rgw_restore_index_tool.py]: sure.

**Review by mkasturi18 — APPROVED** [review]
changes looks good.

**Pull request merged**